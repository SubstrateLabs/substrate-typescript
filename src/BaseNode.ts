import * as Schema from "substrate/Schema";
import { AbstractNode } from "substrate/AbstractNode";

type Props = {
  id?: Schema.Id;
  args?: BaseNode.Args;
};

/**
 * `Node` used in Substrate graph executions.
 */
export class BaseNode extends AbstractNode<BaseNode.Node> implements BaseNode.Node {
  readonly class = "Node" as const;
  extra_args: Object = {};

  constructor(props: Props = {}) {
    super(props?.id);
    this.setArgs(props.args);
  }

  override setArgs(args: BaseNode.Args = {}) {
    Schema.BaseNodeSchema.shape.args.parse(args);
    return super.setArgs(args);
  }
}

export namespace BaseNode {
  /**
   * `Args` are the default inputs given to the `Node` when it executes. These values may be overridden by values coming from parent nodes in the `Graph`.
   */
  export type Args = Schema.BaseNode["args"];

  /**
   * `Node` is the structure this type of node must conform to.
   */
  export type Node = Schema.BaseNode;

  /**
   * `Input` is the required input for this node. It may be fulfilled by the intersection of `Args` and other `Node` outputs that become this node's input.
   */
  export type Input = Schema.BaseNodeInput;

  /**
   * `Output` is what this node will produce after it executes.
   */
  export type Output = unknown;
}
