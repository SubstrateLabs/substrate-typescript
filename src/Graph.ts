import { DiGraph } from "substrate/DiGraph";
import * as Schema from "substrate/Schema";
import { SubstrateError } from "substrate/Error";
import { RemoteCode } from "./RemoteCode";

/**
 * `Graph` is used to construct computation graphs to execute on the Substrate platform.
 *
 * It contains `Node` instances that are connected via an `Edge`. Each `Node` performs some type of
 *
 * A valid `Graph` is a non-empty **directed graph** (eg. contains no cycles).
 */
export class Graph implements Graph.SubstrateGraph {
  /**
   * `initialArgs` are used as initial inputs for root nodes in the graph.
   */
  initialArgs: Graph.InitialArgs;
  /**
   * Internal directed graph structure
   * @internal
   */
  graph: DiGraph;

  /**
   * Returns a new graph that combines all provided graphs.
   */
  static compose(...graphs: [Graph, ...Graph[]]): Graph {
    graphs.forEach((g) => Schema.GraphSchema.parse(g));

    const initialArgs = Object.assign({}, ...graphs.map((g) => g.initialArgs));
    const dag = DiGraph.compose(...graphs.map((graph) => graph.graph));
    return new Graph(initialArgs, dag);
  }

  constructor(initialArgs: Graph.InitialArgs = {}, DAG: DiGraph = new DiGraph()) {
    if (!(DAG instanceof DiGraph)) {
      throw new TypeError(`Expected DiGraph, got ${DAG}`);
    }

    this.graph = DAG;
    this.initialArgs = initialArgs;
  }

  /**
   * Returns a new Graph that includes provided node.
   */
  withNode(node: Graph.SubstrateNode): Graph {
    Schema.NodeSchema.parse(node);

    const g = DiGraph.compose(this.graph);
    g.addNode([node.id, node]);
    return new Graph(this.initialArgs, g);
  }

  /**
   * Returns a new graph that includes provided edge.
   */
  withEdge([from, to, data = {}]: Graph.NewSubstrateEdge): Graph {
    const remoteCode = RemoteCode.tryParse(data);
    const edgeData = remoteCode ? { adapter: remoteCode } : {};

    Schema.EdgeSchema.parse([from, to, edgeData]);

    const g = DiGraph.compose(this.graph);
    g.addNode([from.id, from]);
    g.addNode([to.id, to]);
    g.addEdge([from.id, to.id, edgeData]);
    return new Graph(this.initialArgs, g);
  }

  /**
   * Returns a new graph that includes provided edges.
   */
  withEdges(edges: Graph.NewSubstrateEdge[]): Graph {
    return edges.reduce((graph, edge, _1, _2) => {
      return graph.withEdge(edge);
    }, this as Graph);
  }

  /**
   * Returns the number of nodes in the graph.
   */
  nodeCount() {
    return this.graph.nodes.length;
  }

  /**
   * Returns the number of edges in the graph.
   */
  edgeCount() {
    return this.graph.edges.length;
  }

  get nodes(): Graph.SubstrateNode[] {
    return this.graph.nodes.map(([_nodeId, node]) => node);
  }

  get edges(): Graph.EdgeIds[] {
    return this.graph.edges;
  }

  get initial_args(): Graph.InitialArgs {
    return this.initialArgs;
  }

  validate() {
    if (this.graph.isEmpty()) {
      throw new SubstrateError("Graph is empty");
    }
    if (this.graph.hasCycle()) {
      throw new SubstrateError("Graph contains a cycle");
    }
  }

  toJSON() {
    return Schema.GraphSchema.parse(this);
  }
}

export namespace Graph {
  /**
   * `SubstrateGraph` is the structure that the Substrate Graph API accepts.
   */
  export type SubstrateGraph = Schema.Graph;

  /**
   * `InitialArgs` may be supplied to the `Graph` which are used as initial inputs for root nodes when they are executed.
   */
  export type InitialArgs = Record<string, any>;

  /**
   * `SubstrateNode` represents a vertex within the `Graph` and typically map to a node running a ML model or transformation function.
   */
  export type SubstrateNode = Schema.Node;

  /**
   * `SubstrateEdge` represents the connection between two `Node` items.
   */
  export type SubstrateEdge = Schema.Edge;
  /**
   * `NewSubstrateEdge` represents a new connection between two `Node` items. They are directed from left to right and may also contain `Attributes`.
   */
  export type NewSubstrateEdge = [Graph.SubstrateNode, Graph.SubstrateNode, Function | {}];

  /**
   * `EdgeIds` represents the connection between two `Node` items by `Id`. They are directed from left to right and may also contain `Attributes`.
   */
  export type EdgeIds = Schema.EdgeIds; // NOTE: imo this could use a better name
}
