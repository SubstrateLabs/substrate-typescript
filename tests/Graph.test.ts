import { vi, afterEach, expect, describe, test } from "vitest";
import { Graph } from "substrate/Graph";
import { BaseNode as Node } from "substrate/BaseNode";

describe("Graph", () => {
  const warnMock = vi.spyOn(console, "warn").mockImplementation(() => {});

  afterEach(() => {
    warnMock.mockReset();
  });


  describe(".withNode", () => {
    test("returns a new graph that incldues the node", () => {
      const n = new Node();
      const g = new Graph().withNode(n);
      expect(g.nodes).toContain(n);
    });
  });

  describe(".withEdge", () => {
    test("returns a new graph that incldues the nodes and edges", () => {
      const a = new Node({ id: "a" });
      const b = new Node({ id: "b" });
      const g = new Graph().withEdge([a, b, {}]);
      expect(g.nodes).toEqual([a, b]);
      expect(g.edges).toEqual([["a", "b", {}]]);
    });

    describe("edge with code", () => {
      test("functions are encoded as `AdapterCode`", () => {
        const a = new Node({ id: "a" });
        const b = new Node({ id: "b" });
        const c = new Node({ id: "c" });

        const f = function (x: Node.Output) {
          return { y: x };
        };

        const g = new Graph().withEdge([a, b, f]).withEdge([
          b,
          c,
          function (x: Node.Output): Node.Input {
            return { y: x };
          },
        ]);

        // NOTE: we're matching on snapshots here because the exact whitespace/indentation
        // of the toString() result of the functions can change as the code is reformatted.
        expect(g.edges).toMatchSnapshot();
      });

      test("arrow functions are encoded as `AdapterCode`", () => {
        const a = new Node({ id: "a" });
        const b = new Node({ id: "b" });
        const c = new Node({ id: "c" });

        const f = (x: Node.Output): Node.Input => ({ y: x });

        const g = new Graph()
          .withEdge([a, b, f])
          .withEdge([b, c, (x: Node.Output): Node.Input => ({ y: x })]);

        // NOTE: we're matching on snapshots here because the exact whitespace/indentation
        // of the toString() result of the functions can change as the code is reformatted.
        expect(g.edges).toMatchSnapshot();
      });
    });
  });

  describe(".withEdges", () => {
    test("returns a new graph that incldues the nodes and edges", () => {
      const a = new Node({ id: "a" });
      const b = new Node({ id: "b" });
      const c = new Node({ id: "c" });
      const g = new Graph().withEdges([
        [a, b, {}],
        [b, c, {}],
      ]);
      expect(g.nodes).toEqual([a, b, c]);
      expect(g.edges).toEqual([
        ["a", "b", {}],
        ["b", "c", {}],
      ]);
    });
  });

  describe("examples", () => {
    test("can build linear graphs", () => {
      // a --> b
      // b --> c
      // c --> d
      const a = new Node({ id: "a" });
      const b = new Node({ id: "b" });
      const c = new Node({ id: "c" });
      const d = new Node({ id: "d" });

      const g = new Graph()
        .withEdge([a, b, {}])
        .withEdge([b, c, {}])
        .withEdge([c, d, {}]);

      expect(g.nodeCount()).toEqual(4);
      expect(g.edgeCount()).toEqual(3);
      expect(() => g.toJSON()).not.toThrow();

      const g2 = new Graph().withEdges([
        [a, b, {}],
        [b, c, {}],
        [c, d, {}],
      ]);

      expect(g2.nodeCount()).toEqual(4);
      expect(g2.edgeCount()).toEqual(3);
      expect(() => g2.toJSON()).not.toThrow();
    });

    test("can build graphs with multiple leafs", () => {
      // a --> b
      // b --> c
      // b --> d
      const a = new Node({ id: "a" });
      const b = new Node({ id: "b" });
      const c = new Node({ id: "c" });
      const d = new Node({ id: "d" });

      const g = new Graph()
        .withEdge([a, b, {}])
        .withEdge([b, c, {}])
        .withEdge([b, d, {}]);

      expect(g.nodeCount()).toEqual(4);
      expect(g.edgeCount()).toEqual(3);
      expect(() => g.toJSON()).not.toThrow();

      const g2 = new Graph().withEdges([
        [a, b, {}],
        [b, c, {}],
        [b, d, {}],
      ]);

      expect(g2.nodeCount()).toEqual(4);
      expect(g2.edgeCount()).toEqual(3);
      expect(() => g2.toJSON()).not.toThrow();
    });

    test("can build disconnected graphs", () => {
      // a --> b
      // c --> d
      const a = new Node({ id: "a" });
      const b = new Node({ id: "b" });
      const c = new Node({ id: "c" });
      const d = new Node({ id: "d" });

      const g = new Graph().withEdge([a, b, {}]).withEdge([c, d, {}]);

      expect(g.nodeCount()).toEqual(4);
      expect(g.edgeCount()).toEqual(2);
      expect(() => g.toJSON()).not.toThrow();

      const g2 = new Graph().withEdges([
        [a, b, {}],
        [c, d, {}],
      ]);

      expect(g2.nodeCount()).toEqual(4);
      expect(g2.edgeCount()).toEqual(2);
      expect(() => g2.toJSON()).not.toThrow();
    });

    test("can build branching graphs", () => {
      // a --> b
      // a --> c
      // b --> d
      // c --> d
      // d --> e
      const a = new Node({ id: "a" });
      const b = new Node({ id: "b" });
      const c = new Node({ id: "c" });
      const d = new Node({ id: "d" });
      const e = new Node({ id: "e" });

      const g = new Graph()
        .withEdge([a, b, {}])
        .withEdge([a, c, {}])
        .withEdge([b, d, {}])
        .withEdge([c, e, {}])
        .withEdge([d, e, {}]);

      expect(g.nodeCount()).toEqual(5);
      expect(g.edgeCount()).toEqual(5);
      expect(() => g.toJSON()).not.toThrow();

      const g2 = new Graph().withEdges([
        [a, b, {}],
        [a, c, {}],
        [b, d, {}],
        [c, e, {}],
        [d, e, {}],
      ]);

      expect(g2.nodeCount()).toEqual(5);
      expect(g2.edgeCount()).toEqual(5);
      expect(() => g2.toJSON()).not.toThrow();
    });
  });

  describe("serializing the graph (toJSON)", () => {
    test("when you have provided un specified properties in the graph, they are still serialized", () => {
      // we would like to allow the SDK to be permissive in such a way that a user that is on
      // an older version of this library may override the types specified in the API (via the version of the OpenAPI spec
      // the library is generated with) so that a user can use properties that may be valid, but not contained in
      // an older version. this may be the case if the user for whatever can not or will not update their version of the
      // SDK library.

      const a = new Node().setArgs({ a: 1, b: 2 });
      const b = new Node().setArgs({ c: 1, d: 2 });
      const g = new Graph().withEdge([a, b, { something: "unknown" }]);
      const graphJSON = g.toJSON();

      expect(graphJSON.nodes[0]!.args).toEqual({ a: 1, b: 2 });
      expect(graphJSON.edges[0]![2]).toEqual({ something: "unknown" });
      expect(warnMock).toHaveBeenCalled();
    });
  });
});
