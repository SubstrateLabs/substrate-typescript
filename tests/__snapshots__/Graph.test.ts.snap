// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Graph > .withEdge > edge with code > arrow functions are encoded as \`AdapterCode\` 1`] = `
[
  [
    "a",
    "b",
    {
      "adapter": AdapterCode {
        "code": "(x) => ({ y: x })",
        "runtime": "deno",
      },
    },
  ],
  [
    "b",
    "c",
    {
      "adapter": AdapterCode {
        "code": "(x) => ({ y: x })",
        "runtime": "deno",
      },
    },
  ],
]
`;

exports[`Graph > .withEdge > edge with code > functions are encoded as \`AdapterCode\` 1`] = `
[
  [
    "a",
    "b",
    {
      "adapter": AdapterCode {
        "code": "function(x) {
          return { y: x };
        }",
        "runtime": "deno",
      },
    },
  ],
  [
    "b",
    "c",
    {
      "adapter": AdapterCode {
        "code": "function(x) {
            return { y: x };
          }",
        "runtime": "deno",
      },
    },
  ],
]
`;
